src/vsphere_cpi/spec/unit/cloud/vsphere/vm_placement_selection_pipeline_spec.rb

```diff
@@ -0,0 +1,135 @@
+require 'spec_helper'
+require 'cloud/vsphere/disk_placement_selection_pipeline'
+require 'cloud/vsphere/resources/datastore'
+
+describe VSphereCloud::VmPlacementSelectionPipeline do
+  subject { described_class.new(*criteria) {[placement_1, placement_2]} }
+
+  def fake_placement(cluster)
+    VSphereCloud::VmPlacement.new(cluster: cluster, datastores: cluster.accessible_datastores.values, hosts: nil)
+  end
+
+
+  def fake_datastore(name, free_space: 4096)
+    VSphereCloud::Resources::Datastore.new(
+      name, true, instance_double('VimSdk::Vim::Datastore'), 8192, free_space
+    ).tap do |resource|
+      allow(resource).to receive(:accessible?).and_return(true)
+    end
+  end
+
+  def fake_cluster(name, *args)
+    instance_double('VSphereCloud::Resources::Cluster',
+      :name => name,
+      :free_memory => 8192,
+      :accessible_datastores => args.map{|ds| [ds.name, ds]}.to_h )
+  end
+
+  let(:criteria) { [disk_config: disk_config, req_memory: 1024] }
```

Comment:
Are all of these required at the top level? It seems like there's a lot of rspec magic going on in this spec file. Can we organize it differently to make it a little easier to read?

Reply:
The idea is to simulate a two cluster scenario with two individual datastores and a shared datastore. I declared these at the top to simulate a mini datacenter for all the examples.Should I document with some comments? 

It can be made specific for each example in there. But I prefer to make a large playground(dummy-datacenter) and then write examples with allowing/disallowing properties of various entities.

What do you think?